File: BusTicketChange/BusTicketChange.cpp
/*
https://www.geeksforgeeks.org/problems/lemonade-change/1
*/

class Solution {
  public:
    bool canServe(vector<int> &arr) {
        int countFives=0, countTens=0, countTwenties=0;
        for(auto &vals: arr){
            int curVal=vals;
            if(curVal==5) countFives++;
            else if(curVal==10){
                if(countFives==0) return false;
                else{
                    countFives--;
                    countTens++;
                }
            } else{
                if(countTens>=1){
                    if(countFives==0) return false;
                    else{
                        countFives--;
                        countTens--;
                        countTwenties++;
                    }
                } else{
                    if(countFives<3) return false;
                    else{
                        countFives-=3;
                        countTwenties++;
                    }
                }
            }
        }
        return true;
    }
};

File: BusTicketChange/BusTicketChange.java
/*
https://www.geeksforgeeks.org/problems/lemonade-change/1
*/

class Solution {
    public boolean canServe(int[] arr) {
        int countFiveCoins=0, countTenCoins=0, countTwentyCoins=0;
        for(int coin: arr){
            if(coin==5){
                countFiveCoins++;
            } else if(coin==10){
                if(countFiveCoins==0) return false;
                else{
                    countFiveCoins--;
                    countTenCoins++;
                }
            } else{
                if(countTenCoins>=1){
                    if(countFiveCoins==0) return false;
                    else{
                        countFiveCoins--;
                        countTenCoins--;
                        countTwentyCoins++;
                    }
                } else{
                    if(countFiveCoins<3) return false;
                    else{
                        countFiveCoins-=3;
                        countTwentyCoins++;
                    }
                }
            }
        }
        return true;
    }
}